### ISSUE TITLE
Create SQL-based backend.

### ISSUE THREAD

**Issue by stephenwithav** [vimukthi-git/beanstalkg#33]
Implement `PriorityQueue` interface using PostgreSQL; MySQL is optional for now.

Inspired by: https://github.com/gree/futurama

**Comment by vimukthi-git** [comment]
Nice, go ahead.

**Comment by stephenwithav** [comment]
...and, a better queuing library already exists: https://github.com/bgentry/que-go

I will test that later today.

**Comment by vimukthi-git** [comment]
If it can be made to work compatibly with our queue interface, thats all we need.

**Comment by stephenwithav** [comment]
Indeed.  That's precisely what I am going to test for.

**Comment by stephenwithav** [comment]
`que-go` won't work with our interface, but it is a good starting point.

**Comment by vimukthi-git** [comment]
I think rolling your own is not that hard too. What if you start by replicating our `job` data structure in pg and then do some queries to implement each `queue` interface function as poc? Do you see any problem with that as a starting point?

**Comment by stephenwithav** [comment]
Not at all.  My energy dropped in the early afternoon because I got no sleep the night before last, so I switched from programming to "read everything you can" mode.

I am definitely going to roll my own.  The main benefit of `Que` and `que-go` is that they have optimized the Postgres queries over time, so I will borrow their sql.

**Comment by vimukthi-git** [comment]
Get some sleep, just chilax for a bit and continue. No rush ha ha. 
Yeah its good to read more and weigh all the options. I'm just trying to brain storm. I don't know the best solution either.

**Comment by stephenwithav** [comment]
I spent a few hours this morning writing down what needs to happen to make this work.  I found a much simpler (and more generic) solution, so I'm going to close this issue and open a new one.

**Comment by vimukthi-git** [comment]
I think we should reopen this and try to complete this as it is. It will serve as an example of how to implement a backend.

**Comment by stephenwithav** [comment]
Agreed.

The table design I am considering:

```
CREATE TABLE beanstalkg_jobs (
    id BIGSERIAL PRIMARY KEY,
    uuid UUID,
    tube VARCHAR(255),
    state E_QUEUE_STATUS,
    created_at TIMESTAMP WITH TIME ZONE,
    completed_at TIMESTAMP WITH TIME ZONE,
    ttr BIGINT,
    delay BIGINT,
    priority BIGINT,
    data TEXT)
```

Can you think of anything else that should be tracked in the database?

**Comment by vimukthi-git** [comment]
@stephenwithav this table design makes sense. May be better to store uuid field as string and data as 65536 bytes binary. Also you probably don't need a extra id field as you can use uuid as the primary key field.

**Comment by stephenwithav** [comment]
@vimukthi-git id vs uuid was definitely a debate in my head.  The `beanstalkd` protocol has `max-job-size` is configurable.  Is `beanstalkg` not going to allow that?

**Comment by vimukthi-git** [comment]
@stephenwithav true job size is configurable, my mistake. Then you would have to choose a binary data type that allows you to store arbitrary amount of data.

**Comment by stephenwithav** [comment]
`max-job-size` appears to be a `uint32` in `beanstalkd`, so something akin to `BLOB`.

**Comment by stephenwithav** [comment]
@vimukthi-git  the Postgres wiki [cautions](https://wiki.postgresql.org/wiki/BinaryFilesInDB) against storing binary files.

The `bytea` data type has performance penalties, while large objects can't be inserted unless they exist on the server--then `lo_import` works.

Have you ever needed a binary job in `beanstalkd` or `beanstalkg`?  I can think of cases where that might be useful, but we could recommend using MySQL for those needing binary storage.

`text` has [no size limit](https://www.postgresql.org/docs/current/static/datatype-character.html), so--assuming non-binary jobs--I think it is the best option.  (We could encode/decode as base64, but that's an unneeded performance hit.)

Are you good with `text` and a warning for Postgres?  I can implement the MySQL implementation as well.
